source/portable/os/ota_os_posix.c

```diff
@@ -117,14 +122,30 @@ OtaOsStatus_t Posix_OtaSendEvent( OtaEventContext_t * pEventCtx,
                                   unsigned int timeout )
 {
     OtaOsStatus_t otaOsStatus = OtaOsSuccess;
+    struct timespec ts = { 0 };
+    time_t currentTime = time( NULL );
```

Comment:
MISRA Rule 21.10 does not allow the use of time.h functions because they are easily misunderstood. This use case is one such very frequent mistake. On linux, there isn't really an alternative to time.h, so we should probably add a deviation to this rule if one does not already exist.

The time returned by calls to time() is meant for absolute timing rather than relative timing. 
This is because it may be adjust backwards or forwards at any time at a rate that does not match realtime. For relative timing purposes, clock_gettime with CLOCK_MONOTONIC is the best choice.

Regardless, mq_timedsend takes an absolute timeout, so a call to time() or clock_gettime() is not necessary. Please review the man page.

Reply:
Thanks for suggestion. I will add the deviation for using time.h.

I have two questions about mq_timedsend/mq_timedreceive. 
1. These functions need the absolute time comparing to the current time. How can I get the absolute timeout without calling time() or clock_gettime()?
2. I tried to replace time() with clock_gettime( CLOCK_MONOTONIC, ... ), but I found that mq_timedreceive returns immediately. I can't use CLOCK_MONOTONIC as the input to mq_timedreceive because it uses CLOCK_REALTIME as timer. The return value of CLOCK_REALTIME is larger than CLOCK_MONOTONIC, so it returns immediately. That way, is it still better to use clock_gettime( CLOCK_REALTIME, .. ) than time( NULL )?

Thanks.

Reply:
This is suboptimal though. Embedded systems often lack an RTC due to cost or battery safety concerns and can have wildly varying CLOCK_REALTIME values after boot-up. 

Commonly, systems sync time after bootup with an NTP server. However, if the mq_timedsend operation were to occur at exactly the right time, it could return early. Blocking for longer than expected is not very likely but could still occur if ntp adjusts CLOCK_REALTIME backwards significantly.

We should consider replacing posix message queues with unix domain sockets (AF_UNIX, SOCK_DGRAM) since those can be used with a monotonic timebase and are still quite widely supported on unix-like operating systems.

Reply:
Mqueue supports epoll/poll and select in Linux. So I use poll to detect if send/recv is available to do with timeout 0 in the latest commit.